examples/pytorch/ogb/directional_GSN/main.py

```diff
@@ -0,0 +1,638 @@
+        out_size: int
+            Output dimension of the layer.
+        dropout: float, optional
+            The ratio of units to dropout. No dropout by default.
+            (Default value = 0.)
+        activation: str or callable, optional
+            Activation function to use.
+            (Default value = relu)
+        b_norm: bool, optional
+            Whether to use batch normalization
+            (Default value = False)
+        bias: bool, optional
+            Whether to enable bias in for the linear layer.
+            (Default value = True)
+        init_fn: callable, optional
+            Initialization function to use for the weight of the layer. Default is
+            :math:`\mathcal{U}(-\sqrt{k}, \sqrt{k})` with :math:`k=\frac{1}{ \text{in_size}}`
+            (Default value = None)
+
+    Attributes
+    ----------
+        dropout: int
+            The ratio of units to dropout.
+        b_norm: int
+            Whether to use batch normalization
+        linear: torch.nn.Linear
+            The linear layer
+        activation: the torch.nn.Module
+            The activation layer
+        init_fn: function
+            Initialization function used for the weight of the layer
+        in_size: int
+            Input dimension of the linear layer
+        out_size: int
+            Output dimension of the linear layer
+    """
+
+    def __init__(self, in_size, out_size, activation='relu', dropout=0., b_norm=False, bias=True, init_fn=None,
+                 device='cpu'):
+        super(FCLayer, self).__init__()
+
+        self.__params = locals()
+        del self.__params['__class__']
+        del self.__params['self']
+        self.in_size = in_size
+        self.out_size = out_size
+        self.bias = bias
+        self.linear = nn.Linear(in_size, out_size, bias=bias).to(device)
+        self.dropout = None
+        self.b_norm = None
+        if dropout:
+            self.dropout = nn.Dropout(p=dropout, device=device)
+        if b_norm:
+            self.b_norm = nn.BatchNorm1d(out_size).to(device)
+        self.init_fn = nn.init.xavier_uniform_
+
+        self.reset_parameters()
+
+    def reset_parameters(self, init_fn=None):
+        init_fn = init_fn or self.init_fn
+        if init_fn is not None:
+            init_fn(self.linear.weight, 1 / self.in_size)
+        if self.bias:
+            self.linear.bias.data.zero_()
+
+    def forward(self, x):
+        h = self.linear(x)
+        if self.dropout is not None:
+            h = self.dropout(h)
+        if self.b_norm is not None:
+            if h.shape[1] != self.out_size:
+                h = self.b_norm(h.transpose(1, 2)).transpose(1, 2)
+            else:
+                h = self.b_norm(h)
+        return h
+
+    def __repr__(self):
+        return self.__class__.__name__ + ' (' \
+               + str(self.in_size) + ' -> ' \
+               + str(self.out_size) + ')'
+
+class DGNLayer(nn.Module):
```

Comment:
Can we reuse the built-in `DGNLayer`? https://docs.dgl.ai/generated/dgl.nn.pytorch.conv.DGNConv.html#dgl.nn.pytorch.conv.DGNConv

Reply:
I don't think we can use the built-in `DGNLayer` since the aggregators are different. Also, the message/reduce functions are totally different since subgraph encodings are edge features while eigenvectors are node features.

Reply:
Then is it possible to expand the previous built-in `DGNLayer`? If so, what do we need to add?

Reply:
It's possible to expand the previous built-in `DGNLayer`, we just need to modify the `message` and `reduce_func` functions. Specifically, we need to replace `return {'msg': self.M(f), 'eig_s': edges.src['eig'], 'eig_d': edges.dst['eig']}` with `return {'msg': self.M(f), 'vector_field': edges.data['vector_field']}`  in `message`, and replace `eig_s` and `eig_d` with `vector_field` similarly in `reduce_func`.

Reply:
I see. Perhaps we will leave it for a future work.